File: scripts/open-source/refresh-opensource.py
import json
import requests
import sys
import operator

RAW_DATA_SOURCE = 'github'

def refresh_opensource_data():
    rawRepositories = load_repositories()
    repositories = structure_data(rawRepositories)
    popularity_chart = calculate_popularity(repositories)

    data = {
        'repos': repositories,
        'popularity': popularity_chart
    }

    with open(sys.argv[1], 'w') as f:
        json.dump(data, f, indent = 4, separators = (',', ': '))


def load_repositories():
    if RAW_DATA_SOURCE == 'github':
        response = requests.get('https://api.github.com/users/allegro/repos?per_page=1000')
        return response.json()
    else:
        with open('repositories.json') as repositories_file:
            return json.load(repositories_file)

def structure_data(rawRepositories):
    repositories = {}
    for rawRepo in rawRepositories:
        repo = {
            'name': rawRepo['name'],
            'url': rawRepo['html_url'],
            'stars': rawRepo['stargazers_count'],
            'description': rawRepo['description']
        }
        repositories[repo['name']] = repo
    return repositories

def calculate_popularity(repositories):
    repositories_stars = []
    for name, repo in repositories.items():
        repositories_stars.append((name, repo['stars']))
    return map(lambda t: t[0], sorted(repositories_stars, key = operator.itemgetter(1), reverse = True))

if __name__ == '__main__':
    if len(sys.argv) == 1:
        print "Pass location of output JSON file as program parameter."
    else:
        refresh_opensource_data()

File: scripts/refresh-opensource.js
const fs = require('fs');
const axios = require('axios');

const SOURCE = 'https://api.github.com/users/allegro/repos?per_page=1000';
const TARGET = `../_data/repositories.json`;
const METADATA = {
    hermes: {
        docs: "http://hermes-pubsub.readthedocs.io/en/latest/",
        twitter: "https://twitter.com/hashtag/hermespubsub"
    },
    ralph: {
        docs: "http://ralph-ng.readthedocs.io/en/latest/user/quickstart/"
    },
    'axion-release-plugin': {
        docs: 'http://axion-release-plugin.readthedocs.io/en/latest/'
    },
    vaas: {
        docs: 'http://vaas.readthedocs.io/en/latest/'
    },
    tipboard: {
        docs: 'http://allegro.tech/tipboard/'
    }
};
const FEATURED = {primary: ['vaas', 'hermes', 'ralph'], secondary: ['marathon-consul', 'bigcache']};

axios.get(SOURCE)
    .then(response => response.data)
    .then(repositories => repositories.filter(atLeast10Stars))
    .then(repositories => ({
        popularity: getPopularRepos(repositories),
        repos: repositories.map(render)
    }))
    .then(jobs => {
        fs.writeFile(TARGET, JSON.stringify(jobs, null, 4), err => {
            if (err) return console.log(err);
            console.log(`The file ${TARGET} was saved!`);
        })
    })
    .catch(fail);

function atLeast10Stars(repo) {
    return repo.stargazers_count > 10;
}

function render(repo) {
    return {
        name: repo.name,
        url: repo.html_url,
        stars: repo.stargazers_count,
        description: repo.description,
        featured: {
            primary: FEATURED.primary.includes(repo.name),
            secondary: FEATURED.secondary.includes(repo.name)
        },
        metadata: Object.assign({docs: null, twitter: null, contact: null}, METADATA[repo.name])
    };
}

function getPopularRepos(repositories) {
    return repositories
        .sort((a, b) => b.stargazers_count - a.stargazers_count)
        .map(repo => repo.name);
}

function fail(error) {
    console.error(error);
    process.exit(1)
}
